### PR TITLE
Several modifications, mainly configurable import

### PR DESCRIPTION
**Pull request by jonerikceberio**
[IMP] Cron task
[IMP] Configurable import from csv
[IMP] Http download
[IMP] Leave chunks to-do and relaunch them
[IMP] Field maps and config as object
Configurable import gets the idea from the old ea-import: select diverse order of fields to import in a model. In this way the user can defini his own import without any development

### CODE REVIEW

**Comment by lukebranch** [comment]
@codingforfun ,

Sorry I was not sure where else to ask this question as I can't find an issues page. Thank you for building such a great extension to the connector framework, it looks like it's going to be incredibly useful.

I've managed to get an import done using the standard fruit_catalog.csv using the connector_flow_example_product module as a base.

I've tried to duplicate the module and add some fields, specifically ID to allow mass import of product image(s) to an already existing product in the database, for example:

**export**.product_template_59953

Unfortunately i'm getting the following error message/traceback on those jobs:

```
2015-12-01 07:10:08,053 4202 DEBUG KYAirsoft openerp.addons.connector.queue.worker: <Job 37e2ccd5-5b13-4bc9-b187-d386954cb6e4, priority:10> started
2015-12-01 07:10:08,142 4202 ERROR KYAirsoft openerp.sql_db: bad query: INSERT INTO "product_template" ("id", "track_all", "rating", "list_price", "color", "track_outgoing", "mes_type", "uom_id", "is_features", "uos_coeff", "sale_ok", "purchase_ok", "sale_price_currency_id", "is_arrival", "company_id", "uom_po_id", "type", "track_incoming", "sale_delay", "active", "rental", "purchase_line_warn", "website_sequence", "produce_delay", "categ_id", "website_published", "track_production", "website_size_x", "website_size_y", "multi_image", "sale_line_warn", "create_uid", "write_uid", "create_date", "write_date") VALUES(nextval('product_template_id_seq'), false, 3, '1.00', 0, false, 'fixed', 1, false, '1.000', true, true, 25, false, 1, 1, 'consu', false, 7.0, true, false, 'no-message', 6233, 1.0, 1, false, false, 1, 1, false, 'no-message', 1, 1, (now() at time zone 'UTC'), (now() at time zone 'UTC')) RETURNING id
Traceback (most recent call last):
  File "/opt/odoo/odoo-server/openerp/sql_db.py", line 234, in execute
    res = self._obj.execute(query, params)
IntegrityError: null value in column "name" violates not-null constraint
DETAIL:  Failing row contains (37083, null, null, 1.00, null, 0, null, 1, fixed, 1, null, 2015-12-01 07:10:08.055465, 1.000, 1, t, 1, null, null, 1, null, 1, null, null, null, null, 2015-12-01 07:10:08.055465, t, f, null, null, consu, null, null, f, null, null, 1, 1, null, 6233, f, f, f, null, null, null, null, f, f, 7, f, t, null, null, no-message, no-message, f, 1, 3, 25).

2015-12-01 07:10:08,143 4202 ERROR KYAirsoft openerp.sql_db: bad query: UPDATE "impexp_chunk" SET "state"='failed',"write_uid"=1,"write_date"=(now() at time zone 'UTC') WHERE id IN (8)
Traceback (most recent call last):
  File "/opt/odoo/odoo-server/openerp/sql_db.py", line 234, in execute
    res = self._obj.execute(query, params)
InternalError: current transaction is aborted, commands ignored until end of transaction block

2015-12-01 07:10:08,144 4202 ERROR KYAirsoft openerp.addons.connector.queue.worker: Traceback (most recent call last):
  File "/opt/odoo/custom/oca/connector/connector/queue/worker.py", line 124, in run_job
    job.perform(session)
  File "/opt/odoo/custom/oca/connector/connector/queue/job.py", line 467, in perform
    self.result = self.func(session, *self.args, **self.kwargs)
  File "/opt/odoo/custom/initOS/connector-interfaces/connector_flow/impexp_task.py", line 36, in run_task
    return session.env['impexp.task'].browse(ids).run_task(**kwargs)
  File "/opt/odoo/odoo-server/openerp/api.py", line 239, in wrapper
    return new_api(self, *args, **kwargs)
  File "/opt/odoo/custom/initOS/connector-interfaces/connector_flow/impexp_task.py", line 147, in run_task
    return task_instance.run(config=config, **kwargs)
  File "/opt/odoo/custom/initOS/connector-interfaces/connector_flow/task/abstract_task.py", line 99, in run
    chunk.write({'state': new_state})
  File "/opt/odoo/odoo-server/openerp/api.py", line 239, in wrapper
    return new_api(self, *args, **kwargs)
  File "/opt/odoo/custom/oca/connector/connector/producer.py", line 62, in write
    result = write_original(self, vals)
  File "/opt/odoo/odoo-server/openerp/api.py", line 239, in wrapper
    return new_api(self, *args, **kwargs)
  File "/opt/odoo/odoo-server/openerp/models.py", line 3779, in write
    self._write(old_vals)
  File "/opt/odoo/odoo-server/openerp/api.py", line 239, in wrapper
    return new_api(self, *args, **kwargs)
  File "/opt/odoo/odoo-server/openerp/api.py", line 547, in new_api
    result = method(self._model, cr, uid, self.ids, *args, **kwargs)
  File "/opt/odoo/odoo-server/openerp/models.py", line 3890, in _write
    cr.execute(query, params + (sub_ids,))
  File "/opt/odoo/odoo-server/openerp/sql_db.py", line 158, in wrapper
    return f(self, *args, **kwargs)
  File "/opt/odoo/odoo-server/openerp/sql_db.py", line 234, in execute
    res = self._obj.execute(query, params)
InternalError: current transaction is aborted, commands ignored until end of transaction block
```

Sorry in advance for commenting on a PR, I was not sure where else to ask this question.

**Comment by jonerikceberio** [comment]
It looks like you have not mapped the name in the product, so you are trying to create a product with no name. ¿Could that be it?
Also, not sure if you are using the code from my PR, but if so, it is not prepared for binary fields. However it will be very easy to replicate the functionality in connector_flow_example_product, by adding the retrieval of an image to the _parse_value function.

**Comment by lukebranch** [comment]
@jonerikceberio ,

Thanks for the suggestion, i'll try it out today. Much appreciated!

I'm just using the standard code from this branch:

https://github.com/initOS/connector-interfaces/tree/8.0-connector_flow

**Comment by lukebranch** [comment]
@jonerikceberio ,

Unfortunately It doesn't seem to be working. 

It first spits out this message:

```
openerp.addons.connector_flow.task.csv_import: Deprecated: run at line 62: 'Session(...).create(...)' has been deprecated in favor of 'Session(...).env['model'].create(...)'
```

Then from what I can understand it seems to be looking at the ID and expecting a null value (ie. no values connected to that ID) for example:

```
export.product_template_59953
```

It then finds a bunch of data already connected to that database ID and spits out this: 

```
2015-12-02 05:05:48,852 4202 DEBUG DemoDatabaseName openerp.addons.base.ir.ir_cron: 0.226s (queue.worker, assign_then_enqueue)
2015-12-02 05:05:50,258 4202 ERROR DemoDatabaseName openerp.sql_db: bad query: INSERT INTO "product_template" ("id", "track_all", "rating", "list_price", "color", "track_outgoing", "mes_type", "uom_id", "is_features", "uos_coeff", "sale_ok", "purchase_ok", "sale_price_currency_id", "is_arrival", "company_id", "uom_po_id", "type", "track_incoming", "sale_delay", "active", "rental", "purchase_line_warn", "website_sequence", "produce_delay", "categ_id", "website_published", "track_production", "website_size_x", "website_size_y", "multi_image", "sale_line_warn", "create_uid", "write_uid", "create_date", "write_date") VALUES(nextval('product_template_id_seq'), false, 3, '1.00', 0, false, 'fixed', 1, false, '1.000', true, true, 25, false, 1, 1, 'consu', false, 7.0, true, false, 'no-message', 6237, 1.0, 1, false, false, 1, 1, false, 'no-message', 1, 1, (now() at time zone 'UTC'), (now() at time zone 'UTC')) RETURNING id
Traceback (most recent call last):
  File "/opt/odoo/odoo-server/openerp/sql_db.py", line 234, in execute
    res = self._obj.execute(query, params)
IntegrityError: null value in column "name" violates not-null constraint
DETAIL:  Failing row contains (37106, null, null, 1.00, null, 0, null, 1, fixed, 1, null, 2015-12-02 05:05:48.759032, 1.000, 1, t, 1, null, null, 1, null, 1, null, null, null, null, 2015-12-02 05:05:48.759032, t, f, null, null, consu, null, null, f, null, null, 1, 1, null, 6237, f, f, f, null, null, null, null, f, f, 7, f, t, null, null, no-message, no-message, f, 1, 3, 25).

2015-12-02 05:05:50,260 4202 ERROR DemoDatabaseName openerp.sql_db: bad query: UPDATE "impexp_chunk" SET "state"='failed',"write_uid"=1,"write_date"=(now() at time zone 'UTC') WHERE id IN (21)
Traceback (most recent call last):
  File "/opt/odoo/odoo-server/openerp/sql_db.py", line 234, in execute
    res = self._obj.execute(query, params)
InternalError: current transaction is aborted, commands ignored until end of transaction block

```

and then spits out this traceback:

```
2015-12-02 05:05:51,818 4202 ERROR KYAirsoft openerp.addons.connector.queue.worker: Traceback (most recent call last):
  File "/opt/odoo/custom/oca/connector/connector/queue/worker.py", line 124, in run_job
    job.perform(session)
  File "/opt/odoo/custom/oca/connector/connector/queue/job.py", line 467, in perform
    self.result = self.func(session, *self.args, **self.kwargs)
  File "/opt/odoo/custom/initOS/connector-interfaces/connector_flow/impexp_task.py", line 36, in run_task
    return session.env['impexp.task'].browse(ids).run_task(**kwargs)
  File "/opt/odoo/odoo-server/openerp/api.py", line 239, in wrapper
    return new_api(self, *args, **kwargs)
  File "/opt/odoo/custom/initOS/connector-interfaces/connector_flow/impexp_task.py", line 147, in run_task
    return task_instance.run(config=config, **kwargs)
  File "/opt/odoo/custom/initOS/connector-interfaces/connector_flow/task/abstract_task.py", line 99, in run
    chunk.write({'state': new_state})
  File "/opt/odoo/odoo-server/openerp/api.py", line 239, in wrapper
    return new_api(self, *args, **kwargs)
  File "/opt/odoo/custom/oca/connector/connector/producer.py", line 62, in write
    result = write_original(self, vals)
  File "/opt/odoo/odoo-server/openerp/api.py", line 239, in wrapper
    return new_api(self, *args, **kwargs)
  File "/opt/odoo/odoo-server/openerp/models.py", line 3779, in write
    self._write(old_vals)
  File "/opt/odoo/odoo-server/openerp/api.py", line 239, in wrapper
    return new_api(self, *args, **kwargs)
  File "/opt/odoo/odoo-server/openerp/api.py", line 547, in new_api
    result = method(self._model, cr, uid, self.ids, *args, **kwargs)
  File "/opt/odoo/odoo-server/openerp/models.py", line 3890, in _write
    cr.execute(query, params + (sub_ids,))
  File "/opt/odoo/odoo-server/openerp/sql_db.py", line 158, in wrapper
    return f(self, *args, **kwargs)
  File "/opt/odoo/odoo-server/openerp/sql_db.py", line 234, in execute
    res = self._obj.execute(query, params)
InternalError: current transaction is aborted, commands ignored until end of transaction block
```

I'm going to try and do this on a clean database with now existing database ID's matching the one's i'm importing with and see if that makes a difference.

Do you have any idea if the code in the module needs to change in order to insert values into an existing product (ID)?

The code i'm using is a very slight modification on the example_product module (i've only added 'id' and removed the extra cost price and sale price fields):
```
import urllib2
from base64 import b64encode

from openerp import models, api
from openerp.addons.connector_flow.task.abstract_task \
    import AbstractChunkReadTask


class ProductCatalogImport(AbstractChunkReadTask):
    def read_chunk(self, config=None, chunk_data=None, async=True):
        product_data = {
            'id': chunk_data.get('id'),
            'name': chunk_data.get('name'),
        }
        product_image_url = chunk_data.get('image_URL')
        if product_image_url:
            url_obj = urllib2.urlopen(product_image_url)
            product_data['image'] = b64encode(url_obj.read())
        self.session.env['product.product'].create(product_data)


class ProductCatalogImportTask(models.Model):
    _inherit = 'impexp.task'

    @api.model
    def _get_available_tasks(self):
        return super(ProductCatalogImportTask, self)._get_available_tasks() \
            + [('product_catalog_import', 'Product Catalog Import')]

    def product_catalog_import_class(self):
        return ProductCatalogImport
```

Here's an example of the data i'm importing:

```
id,name,image_URL
__export__.product_template_59953,R5C-AIR #WER5C-27-11 GBBR,http://upload.wikimedia.org/wikipedia/commons/e/ef/Citrus_lime.png
__export__.product_template_59952,R5C-AIR #WER5C-27-10 GBBR,http://upload.wikimedia.org/wikipedia/commons/e/ef/Citrus_lime.png
__export__.product_template_59951,R5C-AIR #WER5C-27-9 GBBR,http://upload.wikimedia.org/wikipedia/commons/e/ef/Citrus_lime.png
__export__.product_template_59950,R5C-AIR #WER5C-27-8 GBBR,http://upload.wikimedia.org/wikipedia/commons/e/ef/Citrus_lime.png
__export__.product_template_59949,R5C-AIR #WER5C-27-7 GBBR,http://upload.wikimedia.org/wikipedia/commons/e/ef/Citrus_lime.png
```

**Comment by lukebranch** [comment]
@jonerikceberio ,

I've figured out how to get everything working for importing new products into a database without an specific database ID, however I can't figure out how to do either of the following:

1) Define an ID for a product (for example export from one database with the XML ID: **export**.product_template_70, and import into a fresh clean database using the same ID)
2) Update product data for an existing XML ID in the database, for example XML ID: **export**.product_template_70 - where this ID matches a product that already exists in the database that I would like to import photos into for example.

Do you have any suggestions on how this might be achieved by modifying this example:

https://github.com/initOS/connector-interfaces/tree/8.0-connector_flow/connector_flow_example_product

I've looked through the connector documentation and I can't figure it out.

**Comment by lukebranch** [comment]
@jonerikceberio ,

I've just cloned your fork and it seems to have some of the extra features I need for updating records rather than just creating new ones (Maps) which is very handy.

You had mentioned:

"Also, not sure if you are using the code from my PR, but if so, it is not prepared for binary fields. However it will be very easy to replicate the functionality in connector_flow_example_product, by adding the retrieval of an image to the _parse_value function."

Is there any way you would be able to help me set this up? I can't figure out how to do this myself and would be more than happy to pay for your services to get it working for me if you like.

Basically I have an existing database with around 10,000 products already imported, they each have their own ID's and I need to only update the pictures using a URL to import the image and assign it to the already existing product in the database.

Unfortunately I can't figure out how to add the retrieval of an image to the _parse_value function as you've described. Any suggestions would be very much appreciated.

**Comment by jonerikceberio** [comment]
Ok, let's see: this "new approach" that you comment is much more coherent with odoo than the previous one. In the previous one I see a clear problem: if you use the orm (and you should), you cannot specify the id. Odoo creates the record and assigns the id through the postgres. That means that it is very difficult to replicate the exact distribution of ids of another db (i.e. jumps in ids due to deletes).
So, if I understand you correctly, you would do a mapping of the products in db2 with the values from the db1. Ok, you can place the name or code as "pk field", so the connector finds the correct counterpart. Once you have that you need to give the image to each record, and that depends of how you have that image. I see two options:
- Url: to have the url to the image in the filestore system. In this case you can use the same way as the product example, as I told you.
- Base64: you have the base64 of the image, and you would have to replicate the odoo process to save them

Does this perspective fit in your needs?

**Comment by lukebranch** [comment]
@jonerikceberio ,

Thank you for your help, I really appreciate it.

What you've described:

"f I understand you correctly, you would do a mapping of the products in db2 with the values from the db1."

is correct. We have a company with a set of 10,000 products and we'd like to use those records in the other database to populate other databases, this allows us to update one and then use the database ID's to update the rest with any changes to those products.

Regarding what you've said here:

"you can place the name or code as "pk field", so the connector finds the correct counterpart."

by 'code' do you mean the internal_reference (ie. product.template.default_code) or product ID (ie. product.template.id)? And by 'name' do you mean the product name (ie. product.template.name)?

And if I understand correctly from your explanation adding the field as a 'PK' field will search for the correct counterpart in the existing database right?

With the image I have them all organised in a CSV file as URL's.

For example:

```
ID,Name,Preis EK,Preis VK,Image URL
70,R5C-AIR  #WER5C-27-11     GBBR,123,123,http://upload.wikimedia.org/wikipedia/commons/e/ef/Citrus_lime.png
69,R5C-AIR  #WER5C-27-10     GBBR,456,456,http://upload.wikimedia.org/wikipedia/commons/e/ef/Citrus_lime.png
68,R5C-AIR  #WER5C-27-9     GBBR,789,789,http://upload.wikimedia.org/wikipedia/commons/e/ef/Citrus_lime.png
67,R5C-AIR  #WER5C-27-8     GBBR,321,321,http://upload.wikimedia.org/wikipedia/commons/e/ef/Citrus_lime.png
66,R5C-AIR  #WER5C-27-7     GBBR,654,654,http://upload.wikimedia.org/wikipedia/commons/e/ef/Citrus_lime.png
```

For this particular case actually what I have done is as follows:

1) Export product ID's from Company A
2) Add Image URL's mapped to product ID's

What I need to do is re-import into that same database and convert the Image URL's to images in the database attached to the original Product ID's if that makes sense.

So I start with products with no images, and I end up with the same products updated with images from image URL's in my CSV.

**Comment by lukebranch** [comment]
@jonerikceberio ,

I seem to be running into this issue now when running a configurable import with data mapping:

```
Traceback (most recent call last):
  File "/opt/odoo/custom/oca/connector/connector/queue/worker.py", line 124, in run_job
    job.perform(session)
  File "/opt/odoo/custom/oca/connector/connector/queue/job.py", line 467, in perform
    self.result = self.func(session, *self.args, **self.kwargs)
  File "/opt/odoo/custom/jonerikceberio/connector-interfaces/connector_flow/impexp_task.py", line 36, in run_task
    return session.env['impexp.task'].browse(ids).run_task(**kwargs)
  File "/opt/odoo/odoo-server/openerp/api.py", line 239, in wrapper
    return new_api(self, *args, **kwargs)
  File "/opt/odoo/custom/jonerikceberio/connector-interfaces/connector_flow/impexp_task.py", line 186, in run_task
    return task_instance.run(config=config, **kwargs)
  File "/opt/odoo/custom/jonerikceberio/connector-interfaces/connector_flow/task/abstract_task.py", line 104, in run
    kwargs['chunk_data'] = simplejson.loads(chunk_data)
  File "/usr/lib/python2.7/dist-packages/simplejson/__init__.py", line 488, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python2.7/dist-packages/simplejson/decoder.py", line 370, in decode
    obj, end = self.raw_decode(s)
  File "/usr/lib/python2.7/dist-packages/simplejson/decoder.py", line 389, in raw_decode
    return self.scan_once(s, idx=_w(s, idx).end())
TypeError: expected string or buffer
```

**Comment by jonerikceberio** [comment]
> And if I understand correctly from your explanation adding the field as a 'PK' field will search for the  correct counterpart in the existing database right?

Correct: first it searches the target model (i.e. product.template) with all the pks, and if it has results it will try to update (write) and if not, it will create the register

Ok, I don't really see if your approach is the best one for a sinchronization between two dbs (plenty of options there), but here are my two cents: in my last commit the module supports urls for binaries:
https://github.com/jonerikceberio/connector-interfaces/commit/6bac591d17c06d1a3386463a1add37e6c92bfb23
I tried it with the example of the connector_flow_example_product and now I am the proud owner of an apple, a kiwi and a lime as products, with their images. So try it and lets see if it works.

And btw, I am also sorry for using @codingforfun 's repo in this way. Hijack is the word that comes to mind!

**Comment by lukebranch** [comment]
@jonerikceberio ,

Thank you for your help, and my humble apologies for commenting on a PR again. I will test your new commit with support for URL's to binaries.

@codingforfun @jonerikceberio ,

Sorry for Hijacking the PR comments for help. This is an amazing tool and thank you to all that have had a hand in building it.

**Comment by lukebranch** [comment]
@jonerikceberio ,

With a fresh clean installation of Ubuntu 14.04 + Odoo your module is still not working for that connector_flow_example_product module.

Using the import CSV: 
fruit_catalog.csv 

and module: 
connector_flow_example_product     (from your jonerikceberio - 8.0-connector_flow branch)

the import process seems to fail with the chunks to products task. I've added a licecap GIF of the process so you can see steps to reproduce. The result is failed chunks and no products imported.



With the initOS 8.0-connector_flow branch the process works without an issue and I get the fruit products with images appear in the catalog:



I won't keep posting issues here if this is unwanted, however if you'd like me to test your fork of the module and provide me information i'd be happy to fork your jonerikceberio fork and create issues on my fork so that it does not get in the way of the PR notes here.

**Comment by jonerikceberio** [comment]
Ok, let's see: You need to have a mapping like this:


Then a task flow with transitions:


And you launch it:


And you have imported 3 products, the 4th one is not viable because of the worng url:


The 3 imported products:


So I cannot know what's failing in your case... you will have to debug it.

**Comment by lukebranch** [comment]
@jonerikceberio ,

Thank you again for your help, I really appreciate it. I will follow the above instructions exactly as you have shown above and debug/troubleshoot on our end until we can reproduce a successful import.

This module is going to be a massive time-saver, and a vital part of my Odoo toolkit once I can get it working.

**Comment by paulius-sladkevicius** [comment]
@codingforfun why it was never merged? Looks like a lot of nice features was added.
